### PR TITLE
Handle samples with no abundances in heatmap

### PR DESCRIPTION
**Pull request by alenajk**
## Status
- [x] **Ready**

## Description

Currently, the custom plots abundances heatmap includes 0s even when the sample in question has no abundances calculated at all for it. For samples that have some abundances, but 0 for other taxa, a 0 value makes sense, but when a sample has no abundances at all, 0 should not be used as a stand-in. We would like to plot columns for samples fitting these criteria (no abundances calculated) as blank/none, and position them to one side of the chart.

image

## Related PRs
- [x] This PR is independent

### CODE REVIEW

File: onecodex/models/collection.py [hunk 1]
```diff
@@ -362,6 +368,8 @@ def _collate_results(self, metric=None, include_host=None):
         self._cached["results"] = df
```
Comment (audy) [inline onecodex/models/collection.py:368]: Suggestion: add a new function `_calculate_collated_results` that does all of the above math but doesn't save anything in cached results. Then, call said method in `_collate_results`. Use new method in `_cluster_by_sample` so that you avoid the cache altogether.

I think maybe we discussed this and realized that it wasn't that easy so apologies if this doesn't make sense.

**Comment by lynchde** [comment]
Just testing locally with some of my own samples (abundances_w_children), and it looks as though we're clustering the samples without abundances with the samples that have 0 for the top_n species in the heatmap plot. I.e. it looks as though we're treating the no-abundance samples as 0% abundance in the clustering call.
(Screenshot without sample names. White bars are samples that don't have abundance estimates. Navy bars are samples that have abundance estimates, but the top 20 species for the collection are 0% for those samples). 
image

Is there any way we can force the non-abundance samples not to cluster amongst the 0% abundance for top_n samples? I.e. force the samples that don't have abundances to be on the right of the plot, and only cluster that samples that do have abundances estimated?

File: onecodex/dataframes.py [hunk 2]
```diff
@@ -52,12 +59,14 @@ def __init__(
...
         ocx_metadata=None,
         ocx_normalized=None,
+        _all_nan_classification_ids=None,
     ):
         self.ocx_rank = ocx_rank
         self.ocx_metric = ocx_metric
         self.ocx_taxonomy = ocx_taxonomy
         self.ocx_metadata = ocx_metadata
         self.ocx_normalized = ocx_normalized
+        self._all_nan_classification_ids = _all_nan_classification_ids
```
Comment (audy) [inline onecodex/dataframes.py:69]: prepend `ocx_` instead of `_`

File: onecodex/analyses.py [hunk 3]
```diff
@@ -306,8 +322,9 @@ def _to_classification_df(
         normalize="auto",
         table_format="wide",
         include_taxa_missing_rank=False,
+        include_nans=False,
```
Comment (audy) [inline onecodex/analyses.py:325]: Make kwarg names consistent with `_to_functional_df` (`filler` & `fill_missing`). I think the default will have to be different though for backwards compatibility.

File: onecodex/distance.py [hunk 4]
```diff
@@ -141,7 +141,7 @@ def unifrac(self, weighted=True, rank=Rank.Auto):
         """
         import skbio.diversity
 
-        df = self.to_df(rank=rank, normalize=self._guess_normalized())
+        df = self.to_df(rank=rank, normalize=self._guess_normalized()).fillna(0)
```
Comment (audy) [inline onecodex/distance.py:144]: You should be able to pass `replace_nans` and `filler` to `to_df`

File: onecodex/models/collection.py [hunk 5]
```diff
@@ -350,7 +353,10 @@ def _collate_results(self, metric=None, include_host=None):
                 if not include_host and d_tax_id in host_tax_ids:
                     continue
 
-                data[c_idx, tax_id_to_idx[d_tax_id]] = d[metric] or 0
+                if metric in [Metric.AbundanceWChildren, Metric.Abundance]:
```
Comment (audy) [inline onecodex/models/collection.py:356]: I don't think you need this check anymore after lines 342-343. The default value should already be `0` for metrics other than `AbundanceWChildren` and `Abundance`

File: onecodex/analyses.py [hunk 6]
```diff
@@ -246,6 +254,14 @@ def _metadata_fetch(self, metadata_fields, label=None):
 
         return magic_metadata, magic_fields
 
+    @property
+    def all_nan_classification_ids(self):
+        if self.__class__.__name__ == "OneCodexAccessor":
+            if self._all_nan_classification_ids is None:
```
Comment (audy) [inline onecodex/analyses.py:260]: Add comment about why this check is necessary

File: onecodex/analyses.py [hunk 7]
```diff
@@ -19,6 +19,14 @@
 )
 
 
+def get_all_nan_classification_ids(df):
```
Comment (audy) [inline onecodex/analyses.py:22]: Prepend with `_` since this isn't being used anywhere else

File: onecodex/viz/_distance.py [hunk 8]
```diff
@@ -42,19 +42,33 @@ def _compute_distance(self, rank, metric):
         return distances
 
     def _cluster_by_sample(
-        self, rank=Rank.Auto, metric=BetaDiversityMetric.BrayCurtis, linkage=Linkage.Average
+        self,
+        all_nan_classification_ids=None,
```
Comment (audy) [inline onecodex/viz/_distance.py:46]: I thin you can get this from `self.all_nan_classification_ids` since that's defined on `AnalysisMixin` which works for both `ClassificationsDataFrame` and `SampleCollection`

File: onecodex/analyses.py [hunk 9]
```diff
@@ -246,6 +254,14 @@ def _metadata_fetch(self, metadata_fields, label=None):
 
         return magic_metadata, magic_fields
 
+    @property
+    def all_nan_classification_ids(self):
+        if self.__class__.__name__ == "OneCodexAccessor":
```
Comment (audy) [inline onecodex/analyses.py:259]: Add a comment that this case happens when `self` is a `ClassificationsDataFrame`

File: onecodex/viz/_heatmap.py [hunk 10]
```diff
@@ -238,6 +251,25 @@ def plot_heatmap(
...
+            if self.__class__.__name__ == "OneCodexAccessor":
+                # the df of a `OneCodexAccessor` does not have NaNs, and we want to display
+                # all_nan_classificaion_id columns as white instead of color corresponding to 0
+                d = d.replace(0.0, np.nan)
+            dropped.append(d)
+            index = df[(df == classification_id).any(axis=1)].index
+            df = df.drop(index)
+
+        df = df.replace(np.nan, 0)
+        for dropped_df in dropped:
```
Comment (audy) [inline onecodex/viz/_heatmap.py:268]: optimization (`pd.concat` is slow): `pd.concat([df, *dropped])`

**Review by audy — CHANGES_REQUESTED** [review]

File: onecodex/dataframes.py [hunk 11]
```diff
@@ -52,12 +59,14 @@ def __init__(
         ocx_taxonomy=None,
         ocx_metadata=None,
         ocx_normalized=None,
+        _all_nan_classification_ids=None,
```
Comment (audy) [inline onecodex/dataframes.py:62]: prepend ocx_ instead of _ for consistency

File: onecodex/models/collection.py (hunk 5 above)
Comment (alenajk) [reply onecodex/models/collection.py:356]: The value in d[metric] would still be possibly `None` though, so as far as I can see we either need this check or to replace what we have with something like
```
if d[metric]:
    data[c_idx, tax_id_to_idx[d_tax_id]] = d[metric]
```
Maybe that's clearer ^ - wdyt?

Comment (audy) [reply onecodex/models/collection.py:356]: oops. You're right. I think `d[metric] or 0` is clearer.

File: onecodex/viz/_distance.py (hunk 8 above)
Comment (alenajk) [reply onecodex/viz/_distance.py:46]: True, but this is used by `plot_distance`, in which we actually want to exclude the samples/classifications entirely from the plot (in a separate PR). Since the behavior is different, and because we actually use more than just the `ids_in_order`, I wanted to keep this as a parameter so that `plot_distance` could _not_ pass anything and keep its current behavior for now. Let me know if you have thoughts though or another idea about handling these two cases.

**Review by audy — APPROVED** [review]

**Review by GSGerritsen — APPROVED** [review]
This seems like a very reasonable way to handle the nan stuff. One suggestion is that it could be nice to put the text from the PR description (or some variant of it) _somewhere_ in the code. It serves as a nice explanation for all the special handling with nans, up to you though!

File: onecodex/viz/_heatmap.py [hunk 12]
```diff
@@ -238,6 +251,24 @@ def plot_heatmap(
...
+                # all_nan_classification_id columns as white instead of color corresponding to 0
+                d = d.replace(0.0, np.nan)
+            dropped.append(d)
+            index = df[(df == classification_id).any(axis=1)].index
+            df = df.drop(index)
+
+        df = df.replace(np.nan, 0)
+        df = pd.concat([df, *dropped])
+
+        assert set(df["Label"].values) == set(labels_in_order)
```
Comment (jairideout) [inline onecodex/viz/_heatmap.py:270]: is this assert needed?

File: onecodex/viz/_heatmap.py [hunk 13]
```diff
@@ -238,6 +251,24 @@ def plot_heatmap(
                 haxis, header=alt.Header(titleOrient="bottom", labelOrient="bottom")
             )
 
+        # Drop all-NaN rows, convert remaining NaNs to 0s, and concat
+        # dropped all-NaN rows to the end
+        dropped = []
+        for classification_id in all_nan_classification_ids:
```
Comment (jairideout) [inline onecodex/viz/_heatmap.py:257]: Since we know which rows we want to move to the end, I think we can accomplish this with less work:

```python
ids_with_data = []
for classification_id in df.index:
    if classification_id not in all_nan_classification_ids:
        ids_with_data.append(classification_id)

df = df.reindex(ids_with_data + all_nan_classification_ids)
```

File: onecodex/analyses.py [hunk 14]
```diff
@@ -246,6 +254,16 @@ def _metadata_fetch(self, metadata_fields, label=None):
 
         return magic_metadata, magic_fields
 
+    @property
+    def all_nan_classification_ids(self):
```
Comment (jairideout) [inline onecodex/analyses.py:258]: Can we make this a "private" property, i.e. `_all_nan_classification_ids`? It's a pretty internal thing that we might not want to commit to having in the public API (yet)

File: tests/test_viz.py [hunk 15]
```diff
@@ -254,6 +254,80 @@ def test_plot_heatmap(ocx, api_data):
...
+        .isnull()
+        .any()
+        == False  # noqa
+    )
+
+    # Sample with all NaNs should be at the end
+    assert chart.encoding.x.sort[-1] == all_nan_sample.filename
+
+
+def test_plot_heatmap_on_df_plots_all_nan_samples_with_nans(ocx, api_data):
```
Comment (jairideout) [inline tests/test_viz.py:294]: what's the difference between this test and the one above it?

File: onecodex/models/collection.py [hunk 16]
```diff
@@ -339,6 +339,9 @@ def _collate_results(self, metric=None, include_host=None):
 
         data = np.zeros((len(self._classifications), len(tax_info.index)), dtype=metric_dtype)
 
+        if metric in [Metric.AbundanceWChildren, Metric.Abundance]:
+            data[:] = np.nan
```
Comment (jairideout) [inline onecodex/models/collection.py:343]: `ndarray.fill()` is a bit faster ([link](https://stackoverflow.com/a/1704853/3776794)):

```python
data.fill(np.nan)
```

The only reason I mention it is because this part of the code is the slowest part when plotting lots of data

File: onecodex/viz/_heatmap.py (hunk 12 above)
Comment (alenajk) [reply onecodex/viz/_heatmap.py:270]: Not strictly - @audy had suggested it to make sure that we aren't ever ending up in a situation where we have a mismatch between the chart and the `labels_in_order`, which would lead to an incorrect plot

File: tests/test_viz.py (hunk 15 above)
Comment (alenajk) [reply tests/test_viz.py:294]: This tests that the functionality works for `plot_heatmap` called on a `OneCodexAccesor` , ie `samples.to_df().ocx.plot_heatmap`. Since `OneCodexAccessor` uses its own version of `_results`, we need to make sure that we're getting the same functionality (by storing `all_nan_classification_ids` on both `SampleCollection`s and `OneCodexAccessor`s) in both cases

File: onecodex/analyses.py (hunk 14 above)
Comment (alenajk) [reply onecodex/analyses.py:258]: Sure, that makes sense (although it will necessitate renaming the attribute on `OneCodexAccessor`s something else besides `_all_nan_classification_ids` to distinguish between the two). But also, how does leaving it as-is commit us to anything?

File: onecodex/viz/_heatmap.py (hunk 13 above)
Comment (alenajk) [reply onecodex/viz/_heatmap.py:257]: We also need to fill the df with 0s for the `None`s that _don't_ correspond to classifications with no abundances calculated. Also, I think we would still need to replace `0.0` with `np.nan` in the case that we are dealing with a `OneCodexAccessor`

File: onecodex/models/collection.py [hunk 17]
```diff
@@ -350,7 +353,10 @@ def _collate_results(self, metric=None, include_host=None):
                 if not include_host and d_tax_id in host_tax_ids:
                     continue
 
-                data[c_idx, tax_id_to_idx[d_tax_id]] = d[metric] or 0
+                if metric in [Metric.AbundanceWChildren, Metric.Abundance]:
+                    data[c_idx, tax_id_to_idx[d_tax_id]] = d[metric]
```
Comment (pkonieczny) [inline onecodex/models/collection.py:357]: Should we use `d[metric] or np.nan` here?

File: onecodex/viz/_distance.py [hunk 18]
```diff
@@ -64,14 +79,18 @@ def _cluster_by_sample(
...
+        import numpy as np
         from scipy.cluster import hierarchy
         from scipy.spatial.distance import squareform
         from sklearn.metrics.pairwise import euclidean_distances
 
-        dist_matrix = euclidean_distances(self._results.T).round(6)
-        clustering = hierarchy.linkage(squareform(dist_matrix), method=linkage)
+        df = self._results.dropna(how="all").replace(np.nan, 0)
+
+        dist_matrix = euclidean_distances(df.T).round(6)
```
Comment (pkonieczny) [inline onecodex/viz/_distance.py:89]: Could we use https://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.nan_euclidean_distances.html#sklearn.metrics.pairwise.nan_euclidean_distances and not fill in 0's here?

**Review by pkonieczny — APPROVED** [review]
I'm not a big fan of `all_nan_classification_id` params, but I do think it's better to merge as is and refactor later on.

File: onecodex/viz/_distance.py (hunk 18 above)
Comment (pkonieczny) [reply onecodex/viz/_distance.py:89]: Scratch that, we do want the distances to 0 calculated

File: onecodex/viz/_heatmap.py (hunk 13 above)
Comment (jairideout) [reply onecodex/viz/_heatmap.py:257]: I'm confused why a classification in `all_nan_classification_ids` has zeros instead of nans (the "all nan" part of the name is kind of misleading). Is there a way to handle that upstream from here so that `all_nan_classification_ids` will always point to classifications that are all nans?

File: onecodex/viz/_heatmap.py [hunk 19]
```diff
@@ -238,6 +251,24 @@ def plot_heatmap(
...
+            d = df[(df == classification_id).any(axis=1)]
```
Comment (jairideout) [inline onecodex/viz/_heatmap.py:258]: To get a specific row in a dataframe by index label:

```python
row = df.loc[classification_id]
```

File: onecodex/viz/_heatmap.py [hunk 20]
```diff
@@ -238,6 +251,24 @@ def plot_heatmap(
...
+            if self.__class__.__name__ == "OneCodexAccessor":
+                # the df of a `OneCodexAccessor` does not have NaNs, and we want to display
+                # all_nan_classification_id columns as white instead of color corresponding to 0
+                d = d.replace(0.0, np.nan)
+            dropped.append(d)
+            index = df[(df == classification_id).any(axis=1)].index
```
Comment (jairideout) [inline onecodex/viz/_heatmap.py:264]: lines 264 and 265 can be replaced with:

```python
df = df.drop(classification_id)
```

**Comment by jairideout** [comment]
@alenajk @audy maybe it'd be simpler to filter out all nan / zeros samples in `to_df()`, like we do for taxa with the `remove_zeros=True` param? We could raise a warning if any samples are removed, and Custom Plots already catches `PlottingWarning`s and displays them to the user. That's the behavior we want for the beta diversity distance plots ([DEV-8400](https://linear.app/onecodex/issue/DEV-8400/beta-diversity-distance-plot-should-not-cluster-samples-that-we-dont)), maybe we could do the same for heatmaps too?

File: onecodex/models/collection.py (hunk 17 above)
Comment (alenajk) [reply onecodex/models/collection.py:357]: I think it shouldn't actually make a difference, but for consistency it is probably better

File: onecodex/viz/_heatmap.py (hunk 19 above)
Comment (alenajk) [reply onecodex/viz/_heatmap.py:258]: I'm getting `KeyError` with that

File: onecodex/viz/_heatmap.py (hunk 20 above)
Comment (alenajk) [reply onecodex/viz/_heatmap.py:264]: This isn't working for me, actually. It leads to a `KeyError`

**Comment by alenajk** [comment]
@jairideout Thanks for the review! Re: `all_nan_classification_ids` sometimes having 0s, it's not really under our control, since even though we have paramerized `to_df` to allow for _not_ filling with 0s, you can call `plot_heatmap` on a `OneCodexAccesor`, and when it calls `to_df` internally, `to_df` references `self._results`, which is just the df you've instantiated the `OneCodexAccessor` with. This also is why we went further back instead of doing filtering in `to_df` or `to_classification_df`.

I can rename for clarity though - perhaps `no_abundances_classification_ids` or `classification_ids_with_no_abundances` would be clearer.

File: onecodex/models/collection.py (hunk 17 above)
Comment (jairideout) [reply onecodex/models/collection.py:357]: Hmm but then `0.0` or `0` abundances will get coerced into `np.nan`. Is that something we want?

File: onecodex/viz/_heatmap.py (hunk 19 above)
Comment (jairideout) [reply onecodex/viz/_heatmap.py:258]: oh my bad, I thought the rows in the dataframe were classification IDs, and the columns were taxa, but that's only the case when `to_df(table_format="wide")` is used (the format is `"long"` here)

File: onecodex/viz/_heatmap.py (hunk 20 above)
Comment (jairideout) [reply onecodex/viz/_heatmap.py:264]: whoops, didn't realize we were using a `long` format df instead of `wide`

File: onecodex/analyses.py (hunk 14 above)
Comment (alenajk) [reply onecodex/analyses.py]: I have renamed it

File: onecodex/models/collection.py (hunk 17 above)
Comment (alenajk) [reply onecodex/models/collection.py:357]: Ah, good point - no, we want to keep 0 and 0.0. I will revert to d[metric] in this case

File: onecodex/dataframes.py [hunk 21]
```diff
@@ -172,7 +172,7 @@ def __init__(self, pandas_obj):
         self._metric = pandas_obj.ocx_metric
         self._rank = pandas_obj.ocx_rank
         self._results = pandas_obj
-        self._all_nan_classification_ids = pandas_obj.ocx_all_nan_classification_ids
+        self.ocx_all_nan_classification_ids = pandas_obj.ocx_all_nan_classification_ids
```
Comment (jairideout) [inline onecodex/dataframes.py:175]: the other attributes on this object aren't prefixed with `ocx_`, is there a reason this one is? the previous name `_all_nan_classification_ids` made sense to me, and it's a "private" name too

Comment (alenajk) [reply onecodex/dataframes.py:175]: It can't be the same name as the property on `AnalysisMixin` - I can put an underscore though

**Review by jairideout — APPROVED** [review]
